elephant/gpfa.py

```diff
@@ -419,24 +237,189 @@ def gpfa(data, bin_size=20*pq.ms, x_dim=3, em_max_iters=500):
+        return self._fit(seqs_train, verbose)
+
+    def _check_training_data(self, data):
+        if len(data) == 0:
+            raise ValueError("`data` cannot be empty")
+        if not isinstance(data[0][0], neo.SpikeTrain):
+            raise ValueError("structure of the data is not correct: 0-axis "
+                             "should be trials, 1-axis neo spike trains "
+                             "and 2-axis spike times")
+
+    def _format_training_data(self, data):
+        seqs = gpfa_util.get_seq(data, self.bin_size)
+        # Remove inactive units based on training set
+        self.has_spikes_bool = (np.hstack(seqs['y']).mean(1) != 0)
+        for seq in seqs:
+            seq['y'] = seq['y'][self.has_spikes_bool, :]
+        return seqs
+
+    def _fit(self, seqs_train, verbose):
+        # Check if training data covariance is full rank
+        y_all = np.hstack(seqs_train['y'])
+        y_dim = y_all.shape[0]
+
+        if np.linalg.matrix_rank(np.cov(y_all)) < y_dim:
+            errmesg = 'Observation covariance matrix is rank deficient.\n' \
+                      'Possible causes: ' \
+                      'repeated units, not enough observations.'
+            raise ValueError(errmesg)
+
+        if verbose:
+            print('Number of training trials: {}'.format(len(seqs_train)))
+            print('Latent space dimensionality: {}'.format(self.x_dim))
+            print('Observation dimensionality: {}'.format(self.has_spikes_bool.sum()))
+
+        # The following does the heavy lifting.
+        self.params_est, self.fit_info = gpfa_core.fit(
+            seq_train=seqs_train,
+            x_dim=self.x_dim,
+            bin_width=self.bin_size.rescale('ms').magnitude,
+            min_var_frac=self.min_var_frac,
+            em_max_iters=self.em_max_iters,
+            em_tol=self.em_tol,
+            tau_init = self.tau_init.rescale('ms').magnitude,
+            eps_init = self.eps_init,
+            freq_ll = self.freq_ll,
+            verbose=verbose)
+
+        self.fit_info['has_spikes_bool'] = self.has_spikes_bool
+        self.fit_info['min_var_frac'] = self.min_var_frac
+        self.fit_info['bin_size'] = self.bin_size
+
+        return self
+
+    def transform(self, data, returned_data=['xorth']):
+        """
+        Apply dimensionality reduction to the given data with the estimated
+        parameters
+
+        Parameters
+        ----------
+        data : list of list of Spiketrain objects
+            The outer list corresponds to trials and the inner list corresponds
+            to the neurons recorded in that trial, such that data[l][n] is the
+            Spiketrain of neuron n in trial l. Note that the number and order
+            of Spiketrains objects per trial must be fixed such that data[l][n]
+            and data[k][n] refer to the same spike generator for any choice of
+            l,k and n.
+
+        Returns
+        -------
+        seqs_train: numpy.recarray
```

Comment:
1. should be `np.recarray` (numpy is not defined in this scope - it's `np`)
2. since there is no explicit `seqs_train` output variable that is returned, it's better to put it that way:
```
Returns
-------
np.recarray
<tab> Description
```